### PR TITLE
Refactor customer ticketing e2e helper

### PR DESCRIPTION
**Pull request by osahtout**
There's 2 version and you can see each by clicking the commits. Choosing either one is simpler so I'll go with the team preference.

### CODE REVIEW

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  cy.get('cx-customer-ticketing-create-dialog').contains('Add New Request');
+}
+
+
+export function fillTicketDetails(ticketDetails: TestTicketDetails){
+  const CATEGORY_SELECT = 0;
+
+  cy.get('cx-customer-ticketing-create-dialog').within(() => {
+    cy.get('.cx-customer-ticket-form-container').should('contain', 'Add New Request');
+    cy.get('textarea').first().type(ticketDetails.subject);
```
Comment (khristian23) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts]: I know this was already there, but I prefer not rely on the position of the input field.

```suggestion
    const SUBJECT_CONTROL = 'subject';
    cy.get(`textarea[formcontrolname="${SUBJECT_CONTROL}"]`).type(ticketDetails.subject);
```

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    cy.contains('.cx-customer-ticket-label', 'Category').get('select').eq(CATEGORY_SELECT).select(ticketDetails.category);
+    cy.get('textarea').last().type(ticketDetails.message);
```
Comment (khristian23) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts]:
```suggestion
    const MESSAGE_CONTROL = 'message';
    cy.get(`textarea[formcontrolname="${MESSAGE_CONTROL}"]`).type(ticketDetails.message);
```

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/view-ticket-details.e2e-spec.ts [hunk 3]
```diff
@@ -14,7 +14,8 @@ describe('ticketing', () => {
         customerTicketing.clickMyAccountMenuOption();
         customerTicketing.clickCustomerSupportMenuOption();
         customerTicketing.verifyTicketListingPageVisit();
-        customerTicketing.clickFirstTicketFromTicketListing();
+        customerTicketing.createTicket({subject: "ticket detilas", message: "ticket details", category: customerTicketing.TestCategory.complaint});
```
Comment (khristian23) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/view-ticket-details.e2e-spec.ts]:
```suggestion
        customerTicketing.createTicket({subject: "ticket details", message: "ticket details", category: customerTicketing.TestCategory.complaint});
```

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          category: TestCategory.complaint,
+        };
+
+        const ticketToSort: TestTicketDetails = {
+          subject: "This will go from bottom of the list to the top ",
+          message: "This will go from bottom of the list to the top ",
+          category: TestCategory.complaint,
+        };
+
+        const numberOfOtherTicket = 2;
```
Comment (khristian23) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]:
```suggestion
        const numberOfOtherTickets = 2;
```

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        customerTicketing.verifyNumberOfPagesBasedOnTotalNumberOfTicket(totalNumberOfTicketCreated);
+
+        customerTicketing.createMultipleTickets(numberOfTicketToCreateForMultipagePagination, testTicketDetails);
+        totalNumberOfTicketCreated += numberOfTicketToCreateForMultipagePagination;
+        customerTicketing.verifyNumberOfPagesBasedOnTotalNumberOfTicket(totalNumberOfTicketCreated);
+      });
+
+      it('should sort ticket listing', () => {
+
+        const fistTicket: TestTicketDetails = {
```
Comment (khristian23) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]:
```suggestion
        const firstTicket: TestTicketDetails = {
```

**Review by khristian23 — CHANGES_REQUESTED** [review]

**Review by khristian23 — APPROVED** [review]

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    cy.get(`textarea[formcontrolname="${SUBJECT_CONTROL}"]`).type(ticketDetails.subject);
+    cy.contains('.cx-customer-ticket-label', 'Category').get('select').eq(CATEGORY_SELECT).select(ticketDetails.category);
+    cy.get(`textarea[formcontrolname="${MESSAGE_CONTROL}"]`).type(ticketDetails.message);
+    cy.get('cx-form-errors').should('not.be.visible');
+  });
+}
+
+
+export function addFile(filename: string){
+  cy.get('cx-file-upload input[type="file"]').attachFile({filePath: '../helpers/customer-ticketing/files-to-upload/' + filename}, { allowEmpty: true });
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/create-tickets.ts]: what's wrong with splitting into multiple lines

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      headerRow.eq(CHANGED_ON_COLUMN).get('td').should('contain', 'Changed On');
+      headerRow.eq(STATUS_COLUMN).get('td').should('contain', 'Status');
+    }
+    else {
+      cy.get('cx-customer-ticketing-list').find('h3').contains("You don't have any request");
+    }
+  });
+}
+
+export function getNumberOfTicket(): number {
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts]: getNumberOfTickets

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  let numberOfTikcets = 0;
+  cy.get('cx-customer-ticketing-list').then(ticketListingElement => {
+    numberOfTikcets = ticketListingElement.find('tbody').length;
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts]: numberofTickets

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+export function verifyStatusOfTicketInList(rowNumber = FIRST_ROW_TICKET_LIST, status = TestStatus.closed) {
+  const row = cy.get('cx-customer-ticketing-list').get('tbody').get('tr').eq(rowNumber);
+  row.get('td').eq(STATUS_COLUMN).contains(status);
+}
+
+export function verifyPaginationDoesNotExist() {
+  cy.get('cx-pagination').should('not.exist');
+}
+
+export function verifyPaginationExists() {
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts]: verifyPaginationExist

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+export function verifyPaginationExistBasedOnTheNumberOfTicketsCreated(numberOfTicketCreated: number) {
+  if(numberOfTicketCreated > MAX_TICKETS_PER_PAGE) {
+    verifyPaginationExists();
+  }
+  else {
+    verifyPaginationDoesNotExist();
+  }
+}
+
+export function verifyNumberOfPagesBasedOnTotalNumberOfTicket(totalNumberOfTicketCreated: number) {
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts]: totalNumberOfTicketsCreated

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  const LEFT_RIGHT_ARROWS = 2;
+  const FIRST_PAGE = 1;
+  const expectedNumberOfPages = Math.floor(totalNumberOfTicketCreated / MAX_TICKETS_PER_PAGE) + FIRST_PAGE + LEFT_RIGHT_ARROWS;
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts]: see above

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+  verifyPaginationExistBasedOnTheNumberOfTicketsCreated(totalNumberOfTicketCreated);
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts]: see above

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
+import { viewportContext } from "../../../helpers/viewport-context";
+import * as customerTicketing from '../../../helpers/customer-ticketing/customer-ticketing';
+import {TestTicketDetails, TestCategory} from '../../../helpers/customer-ticketing/customer-ticketing';
+
+
+
+describe('ticketing', () => {
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]: customer ticketing

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+        const numberOfTicketToCreateForMultipagePagination = 5;
+
+        customerTicketing.loginRegisteredUser();
+        customerTicketing.visitElectronicTicketListingPage();
+        customerTicketing.verifyPaginationDoesNotExist();
+        customerTicketing.createMultipleTickets(numberOfTicketToCreateInitially, testTicketDetails);
+        customerTicketing.verifyPaginationDoesNotExist();
+        customerTicketing.createMultipleTickets(numberOfTicketToCraeteForOnePagePagination, testTicketDetails);
+        customerTicketing.verifyPaginationExists();
+        let totalNumberOfTicketCreated = (numberOfTicketToCreateInitially + numberOfTicketToCraeteForOnePagePagination);
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]: totalNumberOfTicketsCreated

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        customerTicketing.verifyNumberOfPagesBasedOnTotalNumberOfTicket(totalNumberOfTicketCreated);
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]: verifyNumberOfPagesBasedOnTotalNumberOfTickets

**Comment by adolfobee** [comment]
Please setup prettier to maximum 90 - 100 columns

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+      it('should create 6 tickets', () => {
+        const testTicketDetails: TestTicketDetails = {
+          subject: 'Creating tickets for pagination ',
+          message: 'Creating tickets for pagination ',
+          category: TestCategory.complaint,
+        };
+
+        const numberOfTicketToCreateInitially = 5;
+        const numberOfTicketToCraeteForOnePagePagination = 1;
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]: numberOfTicketsToCreateForOnePagination

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+      });
+
+      it('should create 6 tickets', () => {
+        const testTicketDetails: TestTicketDetails = {
+          subject: 'Creating tickets for pagination ',
+          message: 'Creating tickets for pagination ',
+          category: TestCategory.complaint,
+        };
+
+        const numberOfTicketToCreateInitially = 5;
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]: numberOfTicketsToCreateInitially

File: projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        const numberOfTicketToCraeteForOnePagePagination = 1;
+        const numberOfTicketToCreateForMultipagePagination = 5;
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/integration/regression/customer-ticketing/ticket-listing.e2e-spec.ts]: numberOfTicketsToCreateForMultipagePagination

File: projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+export function verifyPaginationDoesNotExist() {
+  cy.get('cx-pagination').should('not.exist');
+}
+
+export function verifyPaginationExist() {
+  cy.get('cx-pagination').should('exist');
+}
+
+export function verifyPaginationExistBasedOnTheNumberOfTicketsCreated(
+  numberOfTicketCreated: number
```
Comment (adolfobee) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/customer-ticketing/customer-ticketing-helpers/ticket-listing.ts:134]: numberofTicketsCreated

**Review by adolfobee — APPROVED** [review]

**Pull request merged**